File: chapter5/src/main/java/chapter5/future/CollectorService.java
package chapter5.future;

import io.vertx.core.AbstractVerticle;
import io.vertx.core.CompositeFuture;
import io.vertx.core.Future;
import io.vertx.core.Promise;
import io.vertx.core.http.HttpServerRequest;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.web.client.HttpResponse;
import io.vertx.ext.web.client.WebClient;
import io.vertx.ext.web.client.predicate.ResponsePredicate;
import io.vertx.ext.web.codec.BodyCodec;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;

public class CollectorService extends AbstractVerticle {

  private final Logger logger = LoggerFactory.getLogger(chapter5.callbacks.CollectorService.class);
  private WebClient webClient;

  @Override
  public void start(Promise<Void> promise) {
    webClient = WebClient.create(vertx);
    vertx.createHttpServer()
      .requestHandler(this::handleRequest)
      .listen(8080)
      .onFailure(promise::fail)
      .onSuccess(ok -> {
        System.out.println("http://localhost:8080/");
        promise.complete();
      });
  }

  private void handleRequest(HttpServerRequest request) {
    CompositeFuture.all(
      fetchTemperature(3000),
      fetchTemperature(3001),
      fetchTemperature(3002))
      .flatMap(this::sendToSnapshot)
      .onSuccess(data -> request.response()
        .putHeader("Content-Type", "application/json")
        .end(data.encode()))
      .onFailure(err -> {
        logger.error("Something went wrong", err);
        request.response().setStatusCode(500).end();
      });
  }

  private Future<JsonObject> sendToSnapshot(CompositeFuture temps) {
    List<JsonObject> tempData = temps.list();
    JsonObject data = new JsonObject()
      .put("data", new JsonArray()
        .add(tempData.get(0))
        .add(tempData.get(1))
        .add(tempData.get(2)));
    return webClient
      .post(4000, "localhost", "/")
      .expect(ResponsePredicate.SC_SUCCESS)
      .sendJson(data)
      .map(response -> data);
  }

  private Future<JsonObject> fetchTemperature(int port) {
    return webClient
      .get(port, "localhost", "/")
      .expect(ResponsePredicate.SC_SUCCESS)
      .as(BodyCodec.jsonObject())
      .send()
      .map(HttpResponse::body);
  }
}

File: chapter5/src/main/kotlin/chapter5/kotlin/coroutines/CollectorService.kt
package chapter5.kotlin.coroutines

import io.vertx.core.http.HttpServerRequest
import io.vertx.core.json.JsonObject
import io.vertx.ext.web.client.WebClient
import io.vertx.ext.web.client.predicate.ResponsePredicate
import io.vertx.ext.web.codec.BodyCodec
import io.vertx.kotlin.core.http.listenAwait
import io.vertx.kotlin.core.json.Json
import io.vertx.kotlin.core.json.array
import io.vertx.kotlin.core.json.json
import io.vertx.kotlin.core.json.obj
import io.vertx.kotlin.coroutines.CoroutineVerticle
import io.vertx.kotlin.ext.web.client.sendAwait
import io.vertx.kotlin.ext.web.client.sendJsonAwait
import kotlinx.coroutines.async
import kotlinx.coroutines.launch
import org.slf4j.LoggerFactory

class CollectorService : CoroutineVerticle() {

  private val logger = LoggerFactory.getLogger(CollectorService::class.java)

  private lateinit var webClient: WebClient

  override suspend fun start() {
    webClient = WebClient.create(vertx)
    vertx.createHttpServer()
      .requestHandler(this::handleRequest)
      .listenAwait(8080)
  }

  private fun handleRequest(request: HttpServerRequest) {
    launch {
      try {
        val t1 = async { fetchTemperature(3000) }
        val t2 = async { fetchTemperature(3001) }
        val t3 = async { fetchTemperature(3002) }

        val array = Json.array(t1.await(), t2.await(), t3.await())
        val json = json { obj("data" to array) }

        sendToSnapshot(json)
        request.response()
          .putHeader("Content-Type", "application/json")
          .end(json.encode())

      } catch (err: Throwable) {
        logger.error("Something went wrong", err)
        request.response().setStatusCode(500).end()
      }
    }
  }

  private suspend fun fetchTemperature(port: Int): JsonObject {
    return webClient
      .get(port, "localhost", "/")
      .expect(ResponsePredicate.SC_SUCCESS)
      .`as`(BodyCodec.jsonObject())
      .sendAwait()
      .body()
  }

  private suspend fun sendToSnapshot(json: JsonObject) {
    webClient
      .post(4000, "localhost", "/")
      .expect(ResponsePredicate.SC_SUCCESS)
      .sendJsonAwait(json)
  }
}
